### PR TITLE
OCPBUGSM-29819: Installation fails after cancel and reset of installation before node reboot

### PR DESCRIPTION
**Pull request by ori-amizur**
Bug 1966545:
This is an edge case that will probably not happen when using with UI.  It happens when cluster starts installing, then installation is cancelled (and possibly reset), and then starts installing again.
Since the cluster preparation for installation may be long, there is a possibility that two go-routines (one per each installation attempt) each preparing the same cluster will run in parallel.
Both of these go-routines are using the same work directory **prefix/cluster-id** for preparation artifacts and may collide with each other.
The solution for it is to let each of these go-routines to use different work directory, so they will not disturb one another.
The work dir is a temporary directory (ioutil.TempDir).

# What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

# How was this code tested?

Please, select one or more if needed:

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested) - Simulated with bash script that starts installation, cancels installation, and starts installation again
- [ ] No tests needed

It's recommended to take a few extra minutes to provide more information about
how this code was tested. Here are some questions that may be worth answering:

- Should this PR be tested by the reviewer? No
- Is this PR relying on CI for an e2e test run? No
- Should this PR be tested in a specific environment? No
- Any logs, screenshots, etc that can help with the review process? 

# Assignees

Please, add one or two reviewers that could help review this PR.

/assign @empovit 

## Checklist

- [x] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by osherdp** [comment]
@ori-amizur if you'll write "Bug 966545: ..." in the PR title it will sync BZ and this PR

**Comment by tsorya** [comment]
don't we need it after installation?

**Comment by ori-amizur** [comment]
> installation

It is erased after upload to S3

https://github.com/openshift/assisted-service/blob/13c2dfc4588e507637d9cb3eed6ce28d701a5690/pkg/generator/generator.go#L77

File: pkg/generator/generator.go [hunk 1]
```diff
@@ -51,9 +52,16 @@ func New(log logrus.FieldLogger, s3Client s3wrapper.API, cfg Config, workDir str
 // GenerateInstallConfig creates install config and ignition files
 func (k *installGenerator) GenerateInstallConfig(ctx context.Context, cluster common.Cluster, cfg []byte, releaseImage string) error {
 	log := logutil.FromContext(ctx, k.log)
-	clusterWorkDir := filepath.Join(k.workDir, cluster.ID.String())
+	err := os.MkdirAll(k.workDir, 0o755)
```
Comment (tsorya) [inline pkg/generator/generator.go:55]: o?

File: pkg/generator/generator.go [hunk 2]
```diff
@@ -51,9 +52,16 @@ func New(log logrus.FieldLogger, s3Client s3wrapper.API, cfg Config, workDir str
...
+	if err != nil {
+		return err
+	}
+	clusterWorkDir, err := ioutil.TempDir(k.workDir, cluster.ID.String()+".")
```
Comment (tsorya) [inline pkg/generator/generator.go:59]: cluster.ID.String()+"." ? /folder/cluster-id. ? maybe better timestamp or something as /folder/cluster-id_temp?

File: pkg/generator/generator.go [hunk 3]
```diff
@@ -51,9 +52,16 @@ func New(log logrus.FieldLogger, s3Client s3wrapper.API, cfg Config, workDir str
...
+	if err != nil {
+		return err
+	}
 	installerCacheDir := filepath.Join(k.workDir, "installercache")
-	err := os.MkdirAll(clusterWorkDir, 0755)
+	err = os.MkdirAll(clusterWorkDir, 0755)
```
Comment (tsorya) [inline pkg/generator/generator.go:64]: i think it is not needed anymore

File: pkg/generator/generator.go (hunk 1 above)
Comment (empovit) [inline pkg/generator/generator.go:55]: What's the meaning of the `o` in `0o755`?

File: pkg/generator/generator.go (hunk 3 above)
Comment (empovit) [reply pkg/generator/generator.go:64]: @tsorya is right, you should have a `clusterWorkDir` by this time, `ioutil.TempDir(k.workDir, cluster.ID.String()+".") created it with the parent directory taken care of by `os.MkdirAll(k.workDir, 0o755)`.

Comment (ori-amizur) [reply pkg/generator/generator.go]: OK

File: pkg/generator/generator.go (hunk 1 above)
Comment (ori-amizur) [reply pkg/generator/generator.go:55]: octal

**Pull request merged**